src/0install-solver/diagnostics.ml

```diff
@@ -27,53 +26,50 @@ struct
       | Restricts of Model.Role.t * Model.impl * Model.restriction list
       | Feed_problem of string
 
-    let pp f = function
-      | UserRequested r -> pf f "User requested %s" (format_restrictions [ r ])
+    let pp = function
+      | UserRequested r -> Pp.verbatimf "User requested %s" (format_restrictions [ r ])
       | ReplacesConflict old ->
-        pf f "Replaces (and therefore conflicts with) %a" format_role old
+        Pp.verbatim "Replaces (and therefore conflicts with) " ++ format_role old
       | ReplacedByConflict replacement ->
-        pf f "Replaced by (and therefore conflicts with) %a" format_role replacement
+        Pp.verbatim "Replaced by (and therefore conflicts with) "
+        ++ format_role replacement
       | Restricts (other_role, impl, r) ->
-        pf
-          f
-          "%a %a requires %s"
-          format_role
-          other_role
-          Model.pp_version
-          impl
-          (format_restrictions r)
-      | Feed_problem msg -> pf f "%s" msg
+        format_role other_role
+        ++ Pp.char ' '
+        ++ Model.pp_version impl
+        ++ Pp.verbatimf " requires %s" (format_restrictions r)
+      | Feed_problem msg -> Pp.verbatim msg
     ;;
   end
 
   (** Represents a single interface in the example (failed) selections produced by the solver.
       It partitions the implementations into good and bad based (initially) on the split from the
       impl_provider. As we explore the example selections, we further filter the candidates. *)
   module Component = struct
-    type rejection_reason =
+    type 'tag rejection_reason =
```

Comment:
```suggestion
    type 'tag rejection_reason =
```
No need to do this, the tag will always be `Stdune.User_message.Style.t`.

Reply:
It's a bit tricky to do, but replace everything with 'tag that you introduced with `Stdune.User_message.Style.t`. We can look at cleaning up that name at another time using our `Import` module, but there will be too many conflicts for now. Sometimes there is a `_` instead of a `'tag` in the interface just so that you are aware. Also any `unit Pp.t` can be taken as `Stdune.User_message.Style.t Pp.t` but that seems to only be for debug stuff.

Reply:
If `'tag` is always `Stdune.User_message.Style.t`, why is it parameterized? We could have a type alias in stdune if it's always the same

Reply:
It's not always the same, sometimes it carries style information like color and italics etc. For the output of the solver none of this is relevant and we should try to keep types simple rather than making them needlessly polymorphic.

Reply:
Should be better now.
I couldn't shadow the long `Stdune.User_message.Style.t` everywhere, so it spills out a little in `dune_pkg`, but that seems fine to me